### ISSUE TITLE
Make iocp 64-bit clean

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5373]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5373|
|Type|defect|
|Created|2011-11-07 19:54:03Z|
|Branch|https://github.com/twisted/twisted/tree/iocp-64bit-5373-2|

iocpsupport is in many places 64-bit unclean. This is due to mismatching declarations. First, HANDLE, SOCKET and ULONG_PTR are really pointer-sized, not int- or long- sized. Second, Cython always defines PY_SSIZE_T_CLEAN, which means functions such as PyObject_AsReadBuffer take a "Py_ssize_t *" parameter, not an "int *". Here is a patch. I've verified that it compiles fine under 64-bit Windows (it reduces the number of compiler warnings a lot), and twisted.internet.test passes fine with "-r iocp".

Attachments:

* [iocp64.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f52/f52225c40fd5c8341f26f539fa01ec364deb2c1c/c683678c3494326f4faefdd04d1bf46ecd7c4272.patch) (259599 bytes) - added by antoine on 2011-11-07 19:54:51Z - 
* [iocp64-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f52/f52225c40fd5c8341f26f539fa01ec364deb2c1c/9eded05fbe18c09bf07949af4c611c1f812570b6.patch) (260546 bytes) - added by antoine on 2011-11-25 20:09:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5373 5373
type__defect defect
reporter__antoine antoine
priority__normal normal
milestone__ 
branch__branches_iocp_64bit_5373_2 branches/iocp-64bit-5373-2
branch_author__antoine antoine
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1320695643000000 1320695643000000
changetime__1322256077000000 1322256077000000
version__None None
owner__jesstess jesstess
cc__antoine cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_1"></a>|@pitrou commented|
|-|-|

(I've left the generated iocpsupport.c changes in the patch. Perhaps you want me to remove them for ease of review)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess commented|
|-|-|

(In [33236]) Branching to 'iocp-64bit-5373'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess commented|
|-|-|

(In [33237]) Apply iocp64.patch by antoine.

refs [#5373](https://github.com/twisted/twisted/issues/5373)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess set owner to @pitrou|
|-|-|

Thanks for the detailed bug report and patch, antoine!

[Here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/iocp-64bit-5373) are the buildbot results for this branch.

Comparing the windows7-64-py2.7 compile warnings from a recent trunk build and this branch:
 * trunk: http://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/1013/steps/compile/logs/warnings%20%2822%29
 * branch: http://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/1017/steps/compile/logs/warnings%20%2813%29

you'll see that a bunch of warnings have gone away, which is great, but 1 warning did get displaced instead of fixed:

```
twisted/internet/iocpreactor/iocpsupport/iocpsupport.c(3206) : warning C4244: '=' : conversion from 'Py_ssize_t' to 'int', possible loss of data
```

is now

```
twisted/internet/iocpreactor/iocpsupport/iocpsupport.c(3438) : warning C4244: 'function' : conversion from 'Py_ssize_t' to 'DWORD', possible loss of data
```

Can you provide a patch against this branch for this warning in wsarecv.pxi:recv? After that I think this will be ready to merge.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_6"></a>|@pitrou commented|
|-|-|

Ok, here is an updated patch. I've only verified it compiles fine (it differs by a single cast from the previous patch).

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

(In [33238]) Branching to 'iocp-64bit-5373-2'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess commented|
|-|-|

(In [33239]) Apply patch iocp64-2.patch by antoine.

refs [#5373](https://github.com/twisted/twisted/issues/5373)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess commented|
|-|-|

Thanks for the new patch, antoine.

In the future, please submit subsequent patches again the branch instead of against trunk. Please also add the 'review' keyword when the ticket is ready to be reviewed again.

[Here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/iocp-64bit-5373-2) are the buildbot results for the new branch. Everything looks good, and the displaced compiler warning is gone:

 * trunk: http://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/1013/steps/compile/logs/warnings%20%2822%29
 * branch: http://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/1018/steps/compile/logs/warnings%20%2812%29

I think this is ready to merge!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess set status to `closed`|
|-|-|

(In [33241]) Merge iocp-64bit-5373-2

Author: antoine
Reviewer: jesstess
Fixes: [#5373](https://github.com/twisted/twisted/issues/5373)

Fix some iocpsupport declarations causing 64-bit compile warnings.
